### PR TITLE
DATAES-407 - Polishing.

### PR DESCRIPTION
**Pull request by xhaggi**
* unify code between ElasticsearchTemplate and ElasticseachRestTemplate
* fix sort by score in ElasticseachRestTemplate
* fix removing an alias in ElasticseachRestTemplate (DATAES-541)
* improve assertion for index name and index type
* reduce dupe code
* add missing generics
* extend ElasticsearchOperations with more methods and use it template
* reuse ObjectMapper in ElasticseachRestTemplate
* organize imports + formatter
* cleanup code + remove unused methods

@mp911de @christophstrobl @akonczak Would you mind taking a look at this polishing commit

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAES).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

**Comment by mp911de** [comment]
Would you mind rebasing the PR before review? Would it also make sense to consider the reactive template into polishing or do you think that's rather something for a separate round of cleanups (i.e. aligning `ElasticsearchRestTemplate` towards the reactive one)?

**Comment by xhaggi** [comment]
>Would you mind rebasing the PR before review? 

Sure and already done.

>Would it also make sense to consider the reactive template into polishing or do you think that's rather >something for a separate round of cleanups (i.e. aligning ElasticsearchRestTemplate towards the >reactive one)?

IMO we should do it in a separate round, with a dedicated Jira issue for that.

File: src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+/**
+ * @author Artur Konczak
+ * @author Mohsin Husen
+ * @author Mark Paluch
+ * @author Mark Janssen
+ * @author Sascha Woo
+ */
+public abstract class AbstractElasticsearchTemplate implements ElasticsearchOperations {
+
+	protected static String[] toArray(List<String> values) {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java]: Let's try to not expose `protected` utility methods as API. How about introducing a package-private utility class for Spring Data ES to ES API conversion. Alternatively, how about default methods on `Query` to retrieve fields, indices, types as array?

File: src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+		Assert.notNull(values, "values must not be null");
+		return values.toArray(new String[values.size()]);
+	}
+
+	protected static MoreLikeThisQueryBuilder.Item[] toArray(MoreLikeThisQueryBuilder.Item... values) {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java]: We should find a better place for this utility method.

File: src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	public ElasticsearchPersistentEntity<?> getPersistentEntityFor(Class<?> type) {
+
+		Assert.notNull(type, "type must not be null!");
+		Assert.isTrue(type.isAnnotationPresent(Document.class), "Unable to identify index name. " + type.getSimpleName()
+				+ " is not a Document. Make sure the document class is annotated with @Document(indexName=\"foo\")");
+
+		return getElasticsearchConverter().getMappingContext().getRequiredPersistentEntity(type);
+	}
+
+	protected List<String> defaultIfEmpty(List<String> strs, Supplier<String> defaultString) {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java]: We should find a better place for this utility method.

File: src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		return CollectionUtils.isEmpty(strs) ? Arrays.asList(defaultString.get()) : strs;
+	}
+
+	protected String defaultIfEmpty(String str, Supplier<String> defaultString) {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java]: We should find a better place for this utility method.

File: src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		return StringUtils.isEmpty(str) ? defaultString.get() : str;
+	}
+
+	protected List<String> extractIds(SearchResponse response) {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java]: We should find a better place for this utility method.

File: src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+		List<String> ids = new ArrayList<>();
+		for (SearchHit hit : response.getHits()) {
+			if (hit != null)
+				ids.add(hit.getId());
+		}
+		return ids;
+	}
+
+	protected <T> Map<String, String> getDefaultSettings(ElasticsearchPersistentEntity<T> persistentEntity) {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java]: How about introducing a package-private `EntityOperations` API that exposes common functionality for entities? All these methods including how to get/set an Id would be subject to moving to such a utility. 

See [MongoDB](https://github.com/spring-projects/spring-data-mongodb/blob/master/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/EntityOperations.java#L183) for further reference.

**Comment by mp911de** [comment]
This one moves quite a few things around. Long-term we want to remove the Transport client so I'm not sure it makes sense to spend the effort needed to get this PR into `master`.

**Comment by xhaggi** [comment]
>This one moves quite a few things around. Long-term we want to remove the Transport client so I'm not sure it makes sense to spend the effort needed to get this PR into master.

@mp911de force-pushed the branch again. I have spent some time dealing with your concerns and it would be really nice if we did not ignore my work. Now most of the stuff is gone from the `AbstractElasticsearchTemplate` and partially moved to `ElasticsearchOperations` class. The changes also address several differences between the two Elasticsearch template classes and unify the code.

File: src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java (hunk 5 above)
Comment (xhaggi) [reply src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java]: For now, I would leave it here and later move this code to the `ElasticsearchRestTemplate`, where it is currently used only once.

**Comment by xhaggi** [comment]
@mp911de After unifying all methods between `ElasticsearchTemplate` and `ElasticsearchRestTemplate` some issues like DATAES-541 are fixed. Should we move the fix to a separate PR?

File: src/test/java/org/springframework/data/elasticsearch/repositories/setting/dynamic/DynamicSettingAndMappingEntityRepositoryTests.java [hunk 7]
```diff
@@ -66,7 +66,7 @@ public void shouldCreateGivenDynamicSettingsForGivenIndex() {
 
 		// then
 		assertThat(elasticsearchTemplate.indexExists(DynamicSettingAndMappingEntity.class)).isTrue();
-		Map<String, Object> map = elasticsearchTemplate.getSetting(DynamicSettingAndMappingEntity.class);
+		Map<String, String> map = elasticsearchTemplate.getSetting(DynamicSettingAndMappingEntity.class);
```
Comment (sothawo) [inline src/test/java/org/springframework/data/elasticsearch/repositories/setting/dynamic/DynamicSettingAndMappingEntityRepositoryTests.java]: see above

File: src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchOperations.java [hunk 8]
```diff
@@ -152,14 +154,14 @@
 	 *
 	 * @param indexName
 	 */
-	Map<String, Object> getSetting(String indexName);
+	Map<String, String> getSetting(String indexName);
```
[the remark this answers is not in the document]
Comment (xhaggi) [reply src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchOperations.java]: I have undone my changes to the value type, but if that should be possible, why does the `settings.get(key)` method return a string? IMO, it's not working as expected.

https://github.com/spring-projects/spring-data-elasticsearch/blob/master/src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplate.java#L1274

https://github.com/spring-projects/spring-data-elasticsearch/blob/master/src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchTemplate.java#L1075

File: src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java [hunk 9]
```diff
@@ -1,28 +1,77 @@
...
- * 
+ *
  * @author Sascha Woo
  */
-public abstract class AbstractElasticsearchTemplate {
+public abstract class AbstractElasticsearchTemplate implements ElasticsearchOperations, ApplicationContextAware {
 
 	private static final Logger LOGGER = LoggerFactory.getLogger(AbstractElasticsearchTemplate.class);
 
-	protected ElasticsearchConverter elasticsearchConverter;
```
[the remark this answers is not in the document]
Comment (xhaggi) [reply src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java]: Undone.

**Comment by xhaggi** [comment]
>But I fear, that if we now merge this into master (and this just about a week before going GA), we first will have a lot of work to get the changes from 4.0.x merged into master after the release and a great part of this will be obsolete in the near future when ES 8 comes out.

I would really appreciate it if these changes could be merged in the master, just because of the fixes and adjustments. @sothawo If you need help rebasing the 4.x branch against master, let me know. I'm sure this isn't as much work as doing all the things done in this PR 😄

**Comment by sothawo** [comment]
Is there anything left from this PR that we should take over into current master? If not, this PR could be deleted.

File: src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplate.java [hunk 10]
```diff
@@ -1474,121 +1409,90 @@ public void refresh(String indexName) {
...
+		return suggest(suggestion, operations.determineIndexName(clazz));
 	}
 
-	private void setPersistentEntityIndexAndType(Query query, Class clazz) {
-		if (query.getIndices().isEmpty()) {
-			query.addIndices(retrieveIndexNameFromPersistentEntity(clazz));
-		}
-		if (query.getTypes().isEmpty()) {
-			query.addTypes(retrieveTypeFromPersistentEntity(clazz));
+	private SearchResponse search(SearchRequest request) {
```
Comment (xhaggi) [inline src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplate.java]: These methods are really handy and reduce the code a lot. Should I rebase this PR or should I open a new one? I could also add an issue for that, if needed.

Comment (sothawo) [reply src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplate.java]: There were quite heavy changes in the internal code structure in DATAES-629 to DATAES-634 especially with the introduction of the `RequestFactory` and the restructuring of the API. I'm not sure if rebasing will really work here

Comment (xhaggi) [reply src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplate.java]: I'll apply the changes that makes sense manually and could force-push to that branch. But if you think it's better to close this PR and open a new one, then it's also good for me.

Comment (sothawo) [reply src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplate.java]: I personally would create a new PR with a new issue

Comment (sothawo) [reply src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplate.java]: and please notice that with DATAES-672 I am just changing the return types of the `SearchOperations` methods from `Page<T>` to `Page<SearchHit<T>>`, same for lists etc. I have it change in my branch for the Template implementations, but I have to fix the Repository implementations to be able to work with that.

**Pull request closed without merging**